### ISSUE TITLE
remove the exception throwing in filterState data fetching, maybe use statusOr instead. 

### ISSUE THREAD

**Issue by stevenzzzz** [envoyproxy/envoy#20792]
*Title*: remove the exception throwing in filterState data fetching, maybe use statusOr instead. 

*Description*:
https://github.com/envoyproxy/envoy/pull/20791 changed the rule of exception throwing in Envoy data-plane, let's remove the exception-throwing code in FilterState.  Presumably return StatusOr or Status to indicate the error instead. 

```
void FilterStateImpl::setData(absl::string_view data_name, std::shared_ptr<Object> data,
                              FilterState::StateType state_type, FilterState::LifeSpan life_span) {
  if (life_span > life_span_) {
    if (hasDataWithNameInternally(data_name)) {
      throw EnvoyException(
          "FilterState::setData<T> called twice with conflicting life_span on the same data_name.");
    }
    maybeCreateParent(ParentAccessMode::ReadWrite);
    parent_->setData(data_name, data, state_type, life_span);
    return;
  }
```

**Comment by stevenzzzz** [comment]
cc @adisuissa @yanavlasov

**Comment by jmarantz** [comment]
I haven't looked at this in detail, but I suspect that if this filter throws it would crash Envoy. We don't have mechanical protection against throwing in the data plane, but we have lint checks against *try* in the data plane. At least in core code.

**Comment by stevenzzzz** [comment]
I can confirm it would throw exception: in our staging environment, when we tried to setData with immutable StateType, it threw and crashed Envoy.

**Comment by jmarantz** [comment]
Yeah -- in the data plane if you throw, it's basically like RELEASE_ASSERT. Unless you also wrote the 'try/catch' in which case the lint checks should block you from submitting unless you take action to subvert those.

**Comment by kyessenov** [comment]
Extensions that are not in upstream repo don't have the same lint checks, and we also saw crashes due to missing try/catch. I think it's a good idea to start using `StatusOr<reference_wrapper<_>>` but that will affect many downstream extensions.
